cluster-test/test-orchestrator/stat_ingester.go

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// This file contains is responsible for ingesting the ringpop stats of the
+// entire cluster. The stats are analyzed in real-time to assess cluster
+// stability and the stats are at the same time written to a file for later
+// analysis.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// The StatIngester is a UDP server that accepts ringpop stats with added
+// timestamps. The StatIngester analyzes the stream so that it knows when the
+// cluster reaches a stable state. It also writes the stream into a file for
+// later analysis.
+type StatIngester struct {
+	// The where the stats are written to.
+	writer io.Writer
+
+	// Protects emptyNodes and wasUnstable
+	sync.Mutex
+
+	// The stat ingester listens for dissemination stats to determine if the
+	// cluster has reached a stable state. When there are no changes being
+	// disseminated by any node, the cluster is said to be stable.
+	// emptyNodes holds track of which nodes are empty and which nodes still
+	// have changes to disseminate.
+	emptyNodes map[string]bool
+
+	// When waiting for the cluster to be stable, we first want to make sure
+	// that the cluster was unstable at some point. This makes sure that any
+	// failure condition we throw at the cluster has taken effect before we
+	// move onto the next failure condition.
+	wasUnstable bool
```

Comment:
Does a name like 'unstable' or 'isUnstable' make more sense? This bool describes the state of the cluster as it currently is observed right?